src/ring_theory/finiteness.lean

```diff
@@ -921,3 +921,49 @@ by simpa [group.fg_iff_monoid.fg] using finite_type_iff_fg
 end monoid_algebra
 
 end monoid_algebra
+
+section vasconcelos
+variables {R : Type*} [comm_ring R] {M : Type*} [add_comm_group M] [module R M] (f : M →ₗ[R] M)
+
+noncomputable theory
+
+/-- The structure of a module `M` over a ring `R` as a module over `polynomial R` when given a
+choice of how `X` acts by choosing a linear map `f : M →ₗ[R] M` -/
+@[simps]
+def module_polynomial_ring_endo : module (polynomial R) M :=
+module.comp_hom M (polynomial.aeval f).to_ring_hom
+
+open polynomial module
+
+/-- A theorem/proof by Vasconcelos, given a finite module `M` over a commutative ring, any
+surjective endomorphism of `M` is also injective. Based on,
+https://math.stackexchange.com/a/239419/31917,
+https://www.ams.org/journals/tran/1969-138-00/S0002-9947-1969-0238839-5/.
+This is similar to `is_noetherian.injective_of_surjective_endomorphism` but only applies in the
+commutative case, but does not use a Noetherian hypothesis. -/
+theorem module.finite.injective_of_surjective_endomorphism [hfg : finite R M]
+  (f_surj : function.surjective f) : function.injective f :=
+begin
+  letI := module_polynomial_ring_endo f,
+  have X_mul : ∀ o, (X : polynomial R) • o = f o,
+  { intro,
+    simp, },
+  have : (⊤ : submodule (polynomial R) M) ≤ ideal.span {X} • ⊤,
+  { intros a ha,
+    obtain ⟨y, rfl⟩ := f_surj a,
+    rw [← X_mul y],
+    exact submodule.smul_mem_smul (ideal.mem_span_singleton.mpr (dvd_refl _)) trivial, },
+  haveI : is_scalar_tower R (polynomial R) M := ⟨λ x y z, by simp⟩,
```

Comment:
I think it would be good to have this as a lemma after `module_polynomial_ring_endo`. Just to indicate it as a natural API piece. I would also move this `haveI` to directly after the `letI` that introduces the module structure.

Reply:
Yeah the issue is I'm not really sure how to state this properly as a lemma.
The type depends on the typeclases inferred, and this depends on f, I've pushed by best attempt but its quite unsatisfying as a lemma to me. Maybe its ok though.